src/chat/chat_params.rs

```diff
@@ -0,0 +1,291 @@
+                            }
+                        }
+                    }
+
+                    frequency_penalty = <MoxinSlider> {
+                        text: "Frequency Penalty"
+                        min: 0.0
+                        max: 1.0
+                    }
+
+                    presence_penalty = <MoxinSlider> {
+                        text: "Presence Penalty"
+                        min: 0.0
+                        max: 1.0
+                    }
+                }
+            }
+        }
+
+        <ChatParamsActions> {
+            padding: {top: 58, left: 25, right: 25}
+        }
+
+        animator: {
+            panel = {
+                default: show,
+                show = {
+                    redraw: true,
+                    from: {all: Forward {duration: 0.3}}
+                    ease: ExpDecay {d1: 0.80, d2: 0.97}
+                    apply: {main_content = { width: 300, draw_bg: {opacity: 1.0} }}
+                }
+                hide = {
+                    redraw: true,
+                    from: {all: Forward {duration: 0.3}}
+                    ease: ExpDecay {d1: 0.80, d2: 0.97}
+                    apply: {main_content = { width: 110, draw_bg: {opacity: 0.0} }}
+                }
+            }
+        }
+    }
+}
+
+#[derive(Live, LiveHook, Widget)]
+pub struct ChatParams {
+    #[deref]
+    view: View,
+
+    #[animator]
+    animator: Animator,
+}
+
+impl Widget for ChatParams {
+    fn handle_event(&mut self, cx: &mut Cx, event: &Event, scope: &mut Scope) {
+        self.view.handle_event(cx, event, scope);
+        self.widget_match_event(cx, event, scope);
+
+        if self.animator_handle_event(cx, event).must_redraw() {
+            self.redraw(cx);
+        }
+    }
+
+    fn draw_walk(&mut self, cx: &mut Cx2d, scope: &mut Scope, walk: Walk) -> DrawStep {
+        let store = scope.data.get::<Store>().unwrap();
+        let ip = &store.chats.inferences_params;
+
+        let temperature = self.slider(id!(temperature));
+        let top_p = self.slider(id!(top_p));
+        let max_tokens = self.slider(id!(max_tokens));
+        let frequency_penalty = self.slider(id!(frequency_penalty));
+        let presence_penalty = self.slider(id!(presence_penalty));
+        let stop = self.text_input(id!(stop));
+        let stream = self.check_box(id!(stream));
+
+        temperature.set_value(ip.temperature.into());
+        top_p.set_value(ip.top_p.into());
+        max_tokens.set_value(ip.max_tokens.into());
+        frequency_penalty.set_value(ip.frequency_penalty.into());
+        presence_penalty.set_value(ip.presence_penalty.into());
+        stop.set_text(&ip.stop);
+
+        // Idk why, but without the check, this binding will not work when the screen is a non empty chat.
+        // I don't get the relation, maybe is something related to the mandatory `ctx` parameter or the
+        // animator.
+        if stream.selected(cx) != ip.stream {
```

Comment:
I presume this is necessary to avoid a situation where you are firing the toggle animation in this `draw_walk`, which cases to trigger redraws, so the animation would be restarted again (without having the change to make any progress).

Maybe it is better to implement this "data binding" relying more in retained mode? I just tested the following and seems to work:

```rs
    fn handle_event(&mut self, cx: &mut Cx, event: &Event, scope: &mut Scope) {
        if let Event::Startup = event {
            let store = scope.data.get::<Store>().unwrap();
            let ip = &store.chats.inferences_params;

            let temperature = self.slider(id!(temperature));
            let top_p = self.slider(id!(top_p));
            let max_tokens = self.slider(id!(max_tokens));
            let frequency_penalty = self.slider(id!(frequency_penalty));
            let presence_penalty = self.slider(id!(presence_penalty));
            let stop = self.text_input(id!(stop));
            let stream = self.check_box(id!(stream));

            temperature.set_value(ip.temperature.into());
            top_p.set_value(ip.top_p.into());
            max_tokens.set_value(ip.max_tokens.into());
            frequency_penalty.set_value(ip.frequency_penalty.into());
            presence_penalty.set_value(ip.presence_penalty.into());
            stop.set_text(&ip.stop);

            stream.set_selected(cx, ip.stream);
        }

        self.view.handle_event(cx, event, scope);
        self.widget_match_event(cx, event, scope);

        ...
    }
    
    fn draw_walk(&mut self, cx: &mut Cx2d, scope: &mut Scope, walk: Walk) -> DrawStep {
        self.view.draw_walk(cx, scope, walk)
    }
```

It is not a situation where the settings values could be changed from other sources, so it feels safe to make the local state source of truth after taking the data from the store at startup time.

Even thinking in the case of having this setting per-conversation, the form data can be easily repopulated from the store just reacting to the existing `ChatHistoryCardAction::ChatSelected` action.

Reply:
> You are firing the toggle animation in this draw_walk, which cases to trigger redraws, so the animation would be restarted again

Yeah, I had that theory for the checkbox widget when I experienced this [other thing](https://github.com/makepad/makepad/pull/485#discussion_r1663268371) in the slider one. Although, I still don't understand why it worked on empty chats. 

I probably had to update that comment at least saying "this is because setting the value also activates the animation as a side effect causing visual issues so we avoid calling it".

> Maybe it is better to implement this "data binding" relying more in retained mode? I just tested the following and seems to work:

I tried the same when I hit the switch problem the first time. But I discarded the commit because of the following:

- We are changing the approach just because one and only one widget is miss behaving.
  - Fixed with a simple if (2 harmless lines).
- We lose proper data binding in exchange of incomplete data binding.
  - Although, as you mention, it works in this current case, now everyone must know to NOT mutate this from the store directly or they would introduce a bug with not immediate visual implications (risky). Yeah we can put a comment saying "hey, this comes from X widget, do not mutate this directly" but we can also not have that risky point of failure on the first place.
- That solution, or the `if` solution. Both are workarounds. Both are just required because of the switch. Both are not required otherwise. And one involves less functional changes and is more robust than the other.
- If we want retained mode, we don't need data binding on the first place, we should directly ask the widget it's data when we need it. Something like one of the following approaches:

![image](https://github.com/moxin-org/moxin/assets/7684329/31f0a7df-d9b5-41d1-9203-a7f91ded951d)

- And I tried doing the previous point, to have no-dirty, no-fragile, robust retained mode UI, but I stopped in the middle because I was like... "hey... I'm seriously changing my whole implementation just because of this dumb switch...?"
  - ...Thanks god I found how to fix the switch.

So, this would be my course of action:
- Let's leave the thing how it's currently done. I don't see any problem with the implementation. `Don't touch it. It works.` said a sage... somewhere... probably...
  - But let's update this comment giving the animation explanation.
- Unless
  - We prefer going fully retained as an agreement.
  - We found an undetected, not easy to resolve issue with this and we MUST go retained, in which case:
    - I would prefer the approach from the diagrams.
    - But... I'm lazy... So I would actually pick your suggestion and add a warning comment in the store and call it a day.